Support a single change stream to listen to all op log events

williamlardier opened the pull request:
- Listen to all collections but the special ones and the mpu shadow buckets.
- Introduce pipeline factories, that will handle all the logic related to pipelines, to simplify other files.
- Introduce the UniqueConnector strategy, to handle the new type of pipeline and allocations in a standard way.
- Update the handling of old connector: the strategies' & their pipeline factory are used to know if an old bucket should be removed, or kept. This means, if we switch between wildcard and non-wildcard pipelines, we ensure a given bucket is never being listened to more than once.

Issue: BB-602

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -261,7 +261,11 @@ class OplogPopulator {
+            if (this._config.numberOfConnectors === 0) {
+                // If the number of connector is set to 0, then we
+                // use a single connector to listen to the whole DB.
+                this._maximumBucketsPerConnector = Infinity;
```
i wonder if it is safer to use this `Infinity` value (vs using `0` as in the config)

semantically, it is more correct indeed ; however since it requires custom handling anyway, maybe having 0 reduces the risk of "missing something" or behaving unexpectedly (e.g. with 0, we will sure it will just fail if we try to go through the regular allocation process, etc...)

francoisferrand reviewed (commented):
I wonder if this is the right approach... This change is trying to fit this single-change-stream approach in the existing code flow (computing which connector should handle the bucket, etc...), while in this approach there is mostly nothing to do : it is really about setting up one connector on system startup, and just call it a day...

(would still need to keep the reconcile loop, to ensure we restart failed connectors, though... and we may need to consider what to do when changing the overall mode)

I don't know if we should change or how, but I think we really need to dig a bit deeper here.

ghost commented:
@francoisferrand shouldn't we also handle "old" connectors on startup? We can detect them and close them to recreate a single connector, or leave them untouched and handle any new connector in a single one. If we only create one connector, I'm not sure this will work for existing deployments.

Edit: as discussed the current code should cover our tests needs. If a customer needs this approach, we will handle it manually, so no need to handle old connectors here.

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

Seragonia committed:
Add wildcard constant as a way to detect unique pipelines

Issue: BB-602

Seragonia committed:
Bind the event in the strategy decorator

This moves the logic out of the OplogPopulator file.

Issue: BB-602

Seragonia committed:
Create the UniqueConnector allocation strategy

- The strategy manages a pipeline that listens to all buckets.

Issue: BB-602

Seragonia committed:
Introduce pipeline factories

- Each facroty manages a specific type of pipeline.
- We use the pipeline in the allocation strategy.
- The Connector now uses the factory from the allocation strategy
  to create the right pipeline.
- The ConnectorsManager now rely on the allocation strategy to
  decide if an old connector should be kept or not, based on the
  old connector config, and the current allocation strategy.

Issue: BB-602

Seragonia committed:
Use the UniqueConnector when the number of connector in conf is 0

Issue: BB-602

Seragonia committed:
Delete porevious connectors not matching the strategy

- This ensures we never listen to a bucket twice
- Handling of "transitions", i.e., to ensure there will be no
  event missed or seen twice, is out of scope, and will be
  covered by a procedure for now.

Issue: BB-602

Seragonia committed:
Update existing tests

Issue: BB-602

Seragonia committed:
Rename getOldConnectors to processOldConnectors

Issue: BB-602

Seragonia committed:
Update unit tests

Issue: BB-602

williamlardier commented:
/wait
/approve

Kerkesni commented on extensions/oplogPopulator/pipeline/WildcardPipelineFactory.js:
```diff
@@ -0,0 +1,46 @@
+                $match: {
+                    'ns.coll': {
+                        $not: {
+                            $regex: `^(${constants.mpuBucketPrefix}|__).*`,
```
Isn't the wildcard element `.*` instead of just `*` ?
Also, do we really need to listen to internal collections ? These are not needed for Backbeat.

welansari commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -134,33 +135,29 @@ class ConnectorsManager extends EventEmitter {
                 const oldConfig = await this._kafkaConnect.getConnectorConfig(connectorName);
                 // extract buckets from old connector config
-                const buckets = this._extractBucketsFromConfig(oldConfig);
+                const buckets = this._allocationStrategy.getOldConnectorBucketList(oldConfig);
+                if (!buckets) {
+                    await this._kafkaConnect.deleteConnector(connectorName);
+                    this._logger.warn('Removed old connector', {
+                        method: 'ConnectorsManager._processOldConnectors',
```
If i understood correctly, when using a single connector to listen to everything, we delete all previous connectors and create a new one with the wildcard element. This might cause us to lose some events or process some events multiple times in case we decide to set a resume time in the config of the connector. Not really sure how we should handle this ?

welansari commented on extensions/oplogPopulator/allocationStrategy/UniqueConnector.js:
```diff
@@ -0,0 +1,44 @@
+        if (!connectors.length) {
+            return null;
+        }
+        return connectors.find(conn => conn.config.pipeline?.includes(constants.wildCardForAllBuckets)) || null;
```
Nit: Should we fail or crash when we don't find the connector ? This should never happen looking at the code. Also if we fall in that case and return `null` a new "wildcard" connector will be created but we're not certain no other connectors are laying around which will be a problem, maybe we should check that ?

williamlardier replied:
So we are supposed to delete any non-wildcard connector if this mode is used, and here we should detect any existing wildcard connector, which should cover all cases... If we do not detect the connector, we need to create it.

francoisferrand commented on extensions/oplogPopulator/allocationStrategy/AllocationStrategy.js:
```diff
@@ -37,6 +39,30 @@ class AllocationStrategy {
+     * @param {Object} oldConfig old configuration
+     * @returns {string[] | null} old configuration if valid
+     */
+    getOldConnectorBucketList(oldConfig) {
+        const bucketList = this.pipelineFactory.extractBucketsFromConfig(oldConfig);
+        if (this.pipelineFactory.isValid(bucketList)) {
+            return bucketList;
+        }
+        return null;
+    }
```
this function does not have to do with the allocation strategy : the "validation" could be moved in the pipeline factory

williamlardier replied:
i moved it to the pipeline factory

francoisferrand commented on extensions/oplogPopulator/allocationStrategy/AllocationStrategy.js:
```diff
@@ -5,9 +5,11 @@ class AllocationStrategy {
      * @param {Logger} params.logger logger object
      */
     constructor(params) {
+        this._pipelineFactory = params.pipelineFactory;
```
is the pipeline factory really *a part of* the strategy?
it seems is is only used as a vessel, to avoid passing an extra parameter to ConnectorManager : but may be cleaner to just keep things separate (lower coupling), and add the parameter where it should be.

williamlardier replied:
Indeed, along with the above comment, I also updated the allocation strategy so it does not hold a reference to the pipeline factory, but we pass it to the ConnectorManager, and it does the same, with simpler code...

Seragonia committed:
Manage the pipeline factory out of the allocation strategy

Issue: BB-602

Seragonia committed:
Remove useless note in test

Issue: BB-602

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-602 contains:


* `8.6.53`

* `9.0.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.54`

* `9.0.0`


Please check the `Fix Version/s` of BB-602, or the target
branch of this pull request.

ghost commented:
/create_integration_branches

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/BB-602` into
`development/8.6`
* [w/9.0/improvement/BB-602](https://github.com/scality/backbeat/compare/development/9.0...w/9.0/improvement/BB-602) will be merged into `development/9.0` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

ghost commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/9.0`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`



Please check the status of the associated issue BB-602.

Goodbye williamlardier.




*The following options are set:* **approve, create_integration_branches**
